Split User Management into Students-only and Admins-only Views

corypride opened the pull request:
## Description of the change
Split Admin user page into a Students and Admins page


- **Related issues**: Link to related issue(s) this PR addresses, if any (use phrases like "fixes #1234" or "closes #1234").

## Screenshot(s)
![image](https://github.com/user-attachments/assets/725e03d4-ddd4-44cc-a650-85eef7e79f14)
![image](https://github.com/user-attachments/assets/819bffe8-1918-4cf3-816c-cfafdf483315)


## Additional context
Closes #382

PThorpe92 commented on backend/src/database/users.go:
```diff
@@ -35,6 +40,31 @@ func (db *DB) GetCurrentUsers(page, itemsPerPage int, facilityId uint, order str
 	return count, users, nil
 }
 
+func (db *DB) GetAdminOrStudentUsers(page, itemsPerPage int, facilityId uint, order string, search string, role string) (int64, []models.User, error) {
```
instead of duplicating this function, can we instead just add a check for the query like
```go
// also make sure it's a pointer to models.User
tx := db.Modal(&models.User{}).("facility_id = ?", facilityId)

switch role {
case "admin":
tx = tx.Where("role = 'admin'")
case "student":
tx = tx.Where("role = 'student'")
}
```
that way you can keep all this in one function :+1:

calisio commented on frontend/src/Pages/Admins.tsx:
```diff
@@ -139,7 +139,7 @@ export default function Users() {
 
     return (
-        <AuthenticatedLayout title="Users" path={['Users']}>
+        <AuthenticatedLayout title="Students" path={['Users']}>
```
make sure you change the path here. It should be "Student Management"

calisio commented on frontend/src/Pages/Students.tsx:
```diff
@@ -0,0 +1,340 @@
+    };
+
+    return (
+        <AuthenticatedLayout title="Students" path={['Users']}>
```
Same thing as above- should be "Student Management"

calisio commented on frontend/src/Pages/Admins.tsx:
```diff
@@ -22,7 +22,7 @@ import { useDebounceValue } from 'usehooks-ts';
 import API from '@/api/api';
 
-export default function Users() {
+export default function Students() {
```
This shouldn't be students- it should be Admin or AdminManagement to make it more specific to what the page is actually doing.

calisio commented on backend/src/database/users.go:
You'll need to fix search so that if you are searching on the admin page, only admin will appear (and vise versa for student page). Right now if you are on the student page and search "Super", Super Admin would be rendered.

calisio commented on frontend/src/Pages/Admins.tsx:
```diff
@@ -175,7 +175,6 @@ export default function Users() {
                                 <span>Name</span>
                             </th>
                             <th>Username</th>
-                            <th>Role</th>
```
While you removed the header, the information is still being rendered in the table (see screenshot inserted below). Make sure you remove the entire column- applies to both pages.
<img width="1508" alt="Screenshot 2024-09-25 at 12 31 43 PM" src="https://github.com/user-attachments/assets/05279ec5-5795-48e4-b898-c4a753902af9">

calisio commented on frontend/src/Pages/Students.tsx:
```diff
@@ -0,0 +1,340 @@
+
+    return (
+        <AuthenticatedLayout title="Students" path={['Users']}>
+            <div className="flex flex-col space-y-6 overflow-x-auto rounded-lg p-4">
```
Would you be able to add in a header here that specifies this is the student management page? Like it shows in the screenshot below
<img width="928" alt="Screenshot 2024-09-25 at 12 36 13 PM" src="https://github.com/user-attachments/assets/a753882d-f57a-4280-9fee-cffbff543760">
Should be a quick fix- just adding in `<h1>Student Management</h1>` in there with

calisio commented on frontend/src/Pages/Students.tsx:
```diff
@@ -0,0 +1,340 @@
+                        />
+                    </div>
+
+                    <div className="tooltip tooltip-left" data-tip="Add User">
```
Can we modify this tooltip to be Add Student? And same for Admin

calisio commented on frontend/src/Pages/Students.tsx:
```diff
@@ -0,0 +1,340 @@
+                                <span>Name</span>
+                            </th>
+                            <th>Username</th>
+                            {/*<th>Role</th>*/}
```
Rather than comment this out, lets just delete this line of code. Not relevant anymore

calisio commented on frontend/src/Pages/Students.tsx:
```diff
@@ -0,0 +1,340 @@
+                )}
+            </div>
+
+            <Modal
+                ref={addUserModal}
+                type={ModalType.Add}
+                item="User"
+                form={<AddUserForm onSuccess={onAddUserSuccess} />}
+            />
```
@PThorpe92 @jtucholski Something to think about - this `AddUserForm` currently has the ability to select the type of user you are adding.. should we remove this to avoid users from adding admin from the student page and vise versa? ie we can just default send the role to the backend for now? In the future when we have varying roles of admin this field would be relevant so not sure if we just want to keep it for now? Let me know your thoughts

PThorpe92 replied:
That is a good point. I think for now the role can just be included to the backend, plus it will default to student if not provided

calisio reviewed (changes requested):
@corypride I left some comments, let me know if you have any questions!

calisio commented:
@corypride Same with this one. Because the git history has diverged a great amount, you'll just have to create a new branch off of main and apply your commits on top of them. Then go back in and submit this PR so it is easier to review and it handles all of the merge conflicts!

calisio commented on backend/src/database/users.go:
```diff
@@ -35,19 +41,28 @@ func (db *DB) GetCurrentUsers(page, itemsPerPage int, facilityId uint, order str
 		Offset(offset).
 		Limit(itemsPerPage).
-		Find(&users).Count(&count).Error; err != nil {
+		Error; err != nil {
```
I think this should still include `.Count(&count)`, so the complete line should be `.Count(&count).Error; err != nil {`

PThorpe92 commented:
Came upon an issue over the weekend, so I fixed it to make the tests pass, and in the process ended up throwing some other fixes in there, sorry to cloud the scope of the PR but it would be difficult to separate those changes now. 

The only thing I changed relevant to this PR is I removed the ability to delete the SuperAdmin account.

corypride commented:
I amended the line of code as suggested, and the tests are failing for me
now.  See error examples below:

2024/09/30 16:34:40
/home/runner/work/UnlockEdv2/UnlockEdv2/backend/src/database/milestones.go:92
sqlite3: constraint failed: UNIQUE constraint failed:
milestones.external_id
443
<https://github.com/UnlockedLabs/UnlockEdv2/actions/runs/11110345263/job/30867827887?pr=388#step:4:444>[0.157ms]
[rows:0] INSERT INTO `milestones`
2024/09/30 16:34:40
/home/runner/work/UnlockEdv2/UnlockEdv2/backend/src/database/users.go:183
record not found
553
<https://github.com/UnlockedLabs/UnlockEdv2/actions/runs/11110345263/job/30867827887?pr=388#step:4:554>[0.075ms]
[rows:0] SELECT * FROM `favorites` WHERE user_id = 4 AND course_id = 4
ORDER BY `favorites`.`id` LIMIT 1
554
<https://github.com/UnlockedLabs/UnlockEdv2/actions/runs/11110345263/job/30867827887?pr=388#step:4:555>===
RUN TestHandleFavoriteProgram/TestAdminCanToggleFavoriteProgramOnOff
555
<https://github.com/UnlockedLabs/UnlockEdv2/actions/runs/11110345263/job/30867827887?pr=388#step:4:556>
556
<https://github.com/UnlockedLabs/UnlockEdv2/actions/runs/11110345263/job/30867827887?pr=388#step:4:557>2024/09/30
16:34:40 /home/runner/work/UnlockEdv2/UnlockEdv2/backend/src/database/users.go:183
record not found
557
<https://github.com/UnlockedLabs/UnlockEdv2/actions/runs/11110345263/job/30867827887?pr=388#step:4:558>
[0.063ms] [rows:0] SELECT * FROM `favorites` WHERE user_id = 1 AND
course_id = 4 ORDER BY `favorites`.`id` LIMIT 1

Steps I took:
* Before I amended the code, I deleted my local branch, and pulled from
origin after Preston's last push to the branch, then i switched to the PR
branch and added one line of code.


On Mon, Sep 30, 2024 at 11:19 AM Preston Thorpe ***@***.***>
wrote:

> Came upon an issue over the weekend, so I fixed it to make the tests pass,
> and in the process ended up throwing some other fixes in there, sorry to
> cloud the scope of the PR but it would be difficult to separate those
> changes now.
>
> The only thing I changed relevant to this PR is I removed the ability to
> delete the SuperAdmin account.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/UnlockedLabs/UnlockEdv2/pull/388#issuecomment-2383641122>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A43WLSCDI7APWNJCCATBLNTZZF2YXAVCNFSM6AAAAABOYV7RUKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDGOBTGY2DCMJSGI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

calisio commented:
Hey @corypride, I put that comment but deleted it right after I published it (spoke too soon & should have waited to hit "comment"), but it must have still appeared as an email. The line actually does include count, just earlier in the code a few lines up.
